sound/soc/sof/core.c

```diff
@@ -358,15 +357,18 @@ static int sof_init_environment(struct snd_sof_dev *sdev)
 		ret = validate_sof_ops(sdev);
 		if (ret < 0) {
 			snd_sof_remove(sdev);
+			snd_sof_remove_late(sdev);
 			return ret;
 		}
 	}
 
+	return 0;
+
 err_machine_check:
-	if (ret) {
-		snd_sof_remove(sdev);
-		sof_ops_free(sdev);
-	}
+	snd_sof_remove(sdev);
+err_sof_probe:
+	snd_sof_remove_late(sdev);
+	sof_ops_free(sdev);
```

Comment:
I think not a problem of this patch, but I think it would be more logical to do this cleanup in sof_probe_continue(). It's a bit out-of-place to call snd_sof_remove() and friends here. But alas, not added by this PR, so can be handled in a follow up.